src/objects/datetime.rs

```diff
@@ -0,0 +1,305 @@
+};
+use ffi::{PyDateTime_DateType, PyDate_Check};
+use ffi::{PyDateTime_Date_GET_DAY, PyDateTime_Date_GET_MONTH, PyDateTime_Date_GET_YEAR};
+use ffi::{PyDateTime_DeltaType, PyDelta_Check};
+use ffi::{
+    PyDateTime_TIME_GET_HOUR, PyDateTime_TIME_GET_MICROSECOND, PyDateTime_TIME_GET_MINUTE,
+    PyDateTime_TIME_GET_SECOND,
+};
+use ffi::{PyDateTime_TZInfoType, PyTZInfo_Check};
+use ffi::{PyDateTime_TimeType, PyTime_Check};
+use object::PyObject;
+use std::os::raw::c_int;
+
+#[cfg(Py_3_6)]
+use ffi::{PyDateTime_DATE_GET_FOLD, PyDateTime_TIME_GET_FOLD};
+
+use instance::Py;
+use python::{Python, ToPyPointer};
+
+// Traits
+pub trait PyDateComponentAccess {
+    fn get_year(&self) -> u32;
+    fn get_month(&self) -> u32;
+    fn get_day(&self) -> u32;
+}
+
+pub trait PyDeltaComponentAccess {
+    fn get_days(&self) -> i32;
+    fn get_seconds(&self) -> i32;
+    fn get_microseconds(&self) -> i32;
+}
+
+pub trait PyTimeComponentAccess {
+    fn get_hour(&self) -> u32;
+    fn get_minute(&self) -> u32;
+    fn get_second(&self) -> u32;
+    fn get_microsecond(&self) -> u32;
+    #[cfg(Py_3_6)]
+    fn get_fold(&self) -> u8;
+}
+
+// datetime.date bindings
+pub struct PyDate(PyObject);
+pyobject_native_type!(PyDate, PyDateTime_DateType, PyDate_Check);
+
+impl PyDate {
+    pub fn new(py: Python, year: u32, month: u32, day: u32) -> PyResult<Py<PyDate>> {
+        let y = year as c_int;
+        let m = month as c_int;
+        let d = day as c_int;
+
+        unsafe {
+            let ptr = PyDateTimeAPI.Date_FromDate.unwrap()(y, m, d, PyDateTimeAPI.DateType);
+            Py::from_owned_ptr_or_err(py, ptr)
+        }
+    }
+
+    pub fn from_timestamp(py: Python, args: &PyObject) -> PyResult<Py<PyDate>> {
+        unsafe {
+            let ptr =
+                PyDateTimeAPI.Date_FromTimestamp.unwrap()(PyDateTimeAPI.DateType, args.as_ptr());
+            Py::from_owned_ptr_or_err(py, ptr)
+        }
+    }
+}
+
+impl PyDateComponentAccess for PyDate {
+    fn get_year(&self) -> u32 {
+        unsafe { PyDateTime_Date_GET_YEAR(self.as_ptr()) as u32 }
+    }
+
+    fn get_month(&self) -> u32 {
+        unsafe { PyDateTime_Date_GET_MONTH(self.as_ptr()) as u32 }
+    }
+
+    fn get_day(&self) -> u32 {
+        unsafe { PyDateTime_Date_GET_DAY(self.as_ptr()) as u32 }
+    }
+}
+
+// datetime.datetime bindings
+pub struct PyDateTime(PyObject);
+pyobject_native_type!(PyDateTime, PyDateTime_DateTimeType, PyDateTime_Check);
+
+impl PyDateTime {
+    pub fn new(
+        py: Python,
+        year: u32,
+        month: u32,
+        day: u32,
+        hour: u32,
+        minute: u32,
+        second: u32,
+        microsecond: u32,
+        tzinfo: &PyObject,
+    ) -> PyResult<Py<PyDateTime>> {
+        let y = year as c_int;
+        let mo = month as c_int;
+        let d = day as c_int;
+        let h = hour as c_int;
+        let mi = minute as c_int;
+        let s = second as c_int;
+        let u = microsecond as c_int;
```

Comment:
Clippy doesn't like this line:

```
warning: 5th binding whose name is just one char
   --> src/objects/datetime.rs:115:13
    |
115 |         let u = microsecond as c_int;
    |             ^
    |
    = note: #[warn(many_single_char_names)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#many_single_char_names
```

The easiest solution is probably inlining those variables